### ISSUE TITLE
regression in integration tests for ubuntu runner images; tests pass locally but fail in GitHub Actions (followup of #6981)

### ISSUE THREAD

**Issue by malandis** [actions/runner-images#7402]
### Description

This is a follow-up to #6981

# Summary of #6981 

Our [organization](https://github.com/momentohq) publishes networked client software in multiple languages. For our python client, we have observed a regression in integration tests when switching to ubuntu-22.04 runners. The integration tests pass consistently when using an ubuntu-20.04 runner, but are consistently flaky and fail when using ubuntu-22.04.

To contrast, when doing local development on ubuntu 22.04, the integration tests consistently pass. We have verified this in multiple environments, both in a desktop environment and on cloud VMs. We have never observed locally the behavior we see on the failed jobs.

@Alexey-Ayupov requested a network capture, and then reported:

>  The only problem I can see is in the DNS protocol communication.

He then escalated to the Azure support team.

# New, related issues

We observe similar timeouts in other languages other than Python. We experience the timeouts in Node.js on both ubuntu-20.04 and ubuntu-22.04; and we also observe timeouts in Rust.

cc @Alexey-Ayupov 

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [X] Ubuntu 20.04
- [X] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Image version and build link

Node.js failed tests for [ubuntu-20.04](https://github.com/momentohq/client-sdk-nodejs/pull/299)
Node.js failed tests for [ubuntu-22.04](https://github.com/momentohq/client-sdk-nodejs/pull/203)
Rust failed tests for [ubuntu-22.04](https://github.com/momentohq/client-sdk-rust/pull/131)

### Is it regression?

Yes. The linked pull requests succeed on macos runners for Node.js or ubuntu-20.04 for rust

### Expected behavior

Integration tests should pass and not time out

### Actual behavior

Integration tests with networked calls fail on some tests, which can change from run to run. See logs in linked pull requests.

### Repro steps

For Node.js, use an ubuntu-20.04 or 22.04 runner image; for Rust an ubunt-20.04 runner image.

Rust network capture [here](https://drive.google.com/file/d/1N4b9xnw_5gMQYI7N-N00cCNk_IsulZlj/view?usp=sharing)

Node.js 20.04 network capture [here](https://drive.google.com/file/d/15bqDbXH4pbWvG6WYy4c6fw6W-8sxRR91/view?usp=sharing)

**Comment by erik-bershel** [comment]
Hey @malandis!
May I ask you to provide repro steps both for Node and Rust.

**Comment by ilia-shipitsin** [comment]
@malandis , timing looks like test met timeout

![image](https://user-images.githubusercontent.com/125650415/231386600-4021228b-8bfa-4aef-afac-5e4600e33bc2.png)


maybe there's some resource bottleneck (probably on runner side), can you try to increase timeout to see whether it will pass or not ?

**Comment by malandis** [comment]
> Hey @malandis! May I ask you to provide repro steps both for Node and Rust?

Node: Open a PR in [this repo](https://github.com/momentohq/client-sdk-nodejs) with the runner image set to ubuntu-20.04 or ubuntu-22.04. See [this PR](https://github.com/momentohq/client-sdk-nodejs/pull/299/files) for an example. Note that the tests pass when we run locally in Ubuntu, remotely on an Ubuntu VM, or on GH with a macos runner.

Rust: open a pull request in [this repo](https://github.com/momentohq/client-sdk-rust). See [this PR](https://github.com/momentohq/client-sdk-rust/pull/131) for reference. The `build-rust` job in the CI workflow intermittently fails.

**Comment by malandis** [comment]
> @malandis , timing looks like test met timeout
> 
> ![image](https://user-images.githubusercontent.com/125650415/231386600-4021228b-8bfa-4aef-afac-5e4600e33bc2.png)
> 
> maybe there's some resource bottleneck (probably on runner side), can you try to increase timeout to see whether it will pass or not ?

@ilia-shipitsin I increased the test timeout first to 60s (from 30s) and then to 120s. Both had the same error. See [here](https://github.com/momentohq/client-sdk-nodejs/pull/430).

**Comment by Alexey-Ayupov** [comment]
Hello @malandis, sorry for the late response. I see a lot of dropped packets toward ipv6 DNS. It looks like your tests are partially trying to get a DNS name using the ipv6 standard. 

To test one of my versions, could you please help me?
For the test purpose, please perform this command:
```sed -i 's/::1 ip6-localhost ip6-loopback/::1     localhost ip6-localhost ip6-loopback/g' /etc/hosts```
Please don't forget to make a backup of your /etc/hosts file.

**Comment by malandis** [comment]
> Please don't forget to make a backup of your /etc/hosts file.

@Alexey-Ayupov just to confirm, you want me to run this locally (there would be no need to backup the hosts file on a test runner, right)?

**Comment by Alexey-Ayupov** [comment]
Hi, you mentioned that tests on your local machine are running successfully. I want to check the behavior of the test after changing the /etc/hosts file. In GitHub runners, we make a loopback for ipv6.

**Comment by malandis** [comment]
> Hello @malandis, sorry for the late response. I see a lot of dropped packets toward ipv6 DNS. It looks like your tests are partially trying to get a DNS name using the ipv6 standard.
> 
> To test one of my versions, could you please help me? For the test purpose, please perform this command: `sed -i 's/::1 ip6-localhost ip6-loopback/::1 localhost ip6-localhost ip6-loopback/g' /etc/hosts` Please don't forget to make a backup of your /etc/hosts file.

I made this change and ran the tests multiple times. They still pass. I also started a fresh Ubuntu 22.04 VM, made the change there, and the tests pass (on multiple invocations).

**Comment by malandis** [comment]
I appreciate the diligence. Thank you!

**Comment by vpolikarpov-akvelon** [comment]
Hey @malandis. Do you observe such behavior still? I'm trying to reproduce this issue, but didn't succeed yet.

**Comment by malandis** [comment]
@vpolikarpov-akvelon yes, we still observe this behavior. The linked pull requests reproduce the behavior by only changing the ubuntu runner image to LTS version. Let me know if you need help running these.

**Comment by cprice404** [comment]
@vpolikarpov-akvelon if it would make it easier for you to test/reproduce, we can give you an auth token that you could set up as a secret on your fork and then it would be possible to run the tests on your PR. Just let us know if there is a good way to get the token to you and we will can send it your way!

**Comment by vpolikarpov-akvelon** [comment]
Hey @cprice404 @malandis. I have some new information regarding this issue.

1. Network dumps analysis demonstrated that in case of failure test suite only connects control endpoint and even doesn't try to start connecting to cache endpoint. Looks like something goes wrong on communication with control endpoint. Unfortunately, it takes place over TLS so I can't analyze it.

2. I tried to execute tests in Azure on virtual machine created from runner image. The issue was there until I assigned public IP address to the network interface of the machine. With public IP address everything works fine.

Taking these points together I would assume that either something blocks requests on the API's side or communication breaks somewhere in Azure or AWS network. Could you please check configuration of your infrastructure in AWS (especially in terms of network security) and verify all the changes made around the moment when you started observing this issue. Also I would like you to examine logs of your service backend: it may be useful to know how those failures look like from backend's point of view (there should be anything in logs since test suite connects and communicates with control endpoint).

**Comment by vpolikarpov-akvelon** [comment]
Hey @cprice404 @malandis. Did you manage to find something that may negatively affect communication with control endpoint?

**Comment by cprice404** [comment]
Hi @vpolikarpov-akvelon thanks so much for looking into this.

We will look at the logs more today and see if anything shows up but in the past this has appeared to be related to DNS (see https://github.com/actions/runner-images/issues/6981#issuecomment-1415360651). If it was a server-side issue with the control endpoint then I would expect the same issue to occur when running the tests via MacOS?

> verify all the changes made around the moment when you started observing this issue.

For most of our repos, the moment we started observing it is when `ubuntu-latest` on github switched from 20.04 to 22.04. It may be that the nodejs one is not the best one to use to diagnose, since it is the only one that fails on both 20.04 and 22.04. The rust one that @malandis linked above may be more representative and smaller in scope.

We can also reproduce it in either of these repos:
https://github.com/momentohq/client-sdk-go
https://github.com/momentohq/client-sdk-python

> I tried to execute tests in Azure on virtual machine created from runner image. The issue was there until I assigned public IP address to the network interface of the machine. With public IP address everything works fine.

This is really interesting. Do you know how the DNS setup changes between those two configurations?

**Comment by vpolikarpov-akvelon** [comment]
I really doubt that it's related with DNS. There is negative response to `AAAA` query in network trace indeed, but there is also positive response to `A` query with the same name, that means that it was resolved to IPv4 address successfully.

Also I wouldn't rely on the fact that everything works fine on macOS runners. They live in environment that differs a lot from the one where Ubuntu and Windows runners are.

**Comment by cprice404** [comment]
> They live in environment that differs a lot from the one where Ubuntu and Windows runners are.

This is one of the data points that makes us think that it must be an environmental issue :)

For example, our python SDK tests currently work fine on github on ubuntu 20.04, but fail on 22.04. Would it be helpful for us to create a PR to reproduce that failure again to show another example?

**Comment by vpolikarpov-akvelon** [comment]
> Would it be helpful for us to create a PR to reproduce that failure again to show another example?

Yeah, I think so. We are in a dead end now, so any information may be useful.

**Comment by cprice404** [comment]
Done, here's a failed run of the python version:

https://github.com/momentohq/client-sdk-python/actions/runs/5591096534/jobs/10221706048?pr=357

**Comment by cprice404** [comment]
This is probably the most representative run of what we frequently see with 22.04 that we do not see with 20.04:

https://github.com/momentohq/client-sdk-python/actions/runs/5591096534/jobs/10222519393

**Comment by vpolikarpov-akvelon** [comment]
@cprice404

I tried running tests in python repo and caught some strange behavior: when I run all tests then `test_create_list_revoke_signing_keys` fails often, but when I run only that one it always passes. What may be the reason for that?

**Comment by cprice404** [comment]
I will look into that, I don't have a good guess off of the top of my head but it's possible that that one is some kind of flakiness in the definition of the test itself. if you comment out that test are you able to reproduce the other failures that I linked in that last link that I posted above?

I would be happy to update that PR to comment out that signing_key test if it would help.

**Comment by vpolikarpov-akvelon** [comment]
I tried to switch signing_key test off and everything started working flawlessly. It's a bit strange that this test fails when is run among others, but always succeeds alone. Anyway it looks like there is something wrong with it. May it conflict with other tests somehow? And may this conflict be platform-dependent? What to you think?

**Comment by vpolikarpov-akvelon** [comment]
I experimented a bit more. Here is interesting observation. Test `create_list_revoke_signing_keys` never fails alone, but it's enough to run it after `list_caches_succeeds` and `list_caches_succeeds_even_if_cred_provider_has_been_printed` to get a failure. But it doesn't fail on the first run, one should restart workflow several times to see it failing. And starting from that moment, it never succeeds unless is given some time to "rest". Also, first failure usually appears as connection reset, while others as remote host unavailability. It indeed looks like something wrong either with the network or with the backend service - like tests app is blocked for some reason by some kind of firewall or filter.

I know, if it's network rules to blame then you may wonder why it works on Ubuntu 20. Well, it may be related with external IP addresses of runner, for example. A crazy assumption: IP addresses of Ubuntu 22 runner addresses may be trusted less then IP addresses of Ubuntu 20 runners (on any network node that is in on the way from GitHub runners net to your backend net). I doubt that we can really check it somehow. I only may suggest you to verify network security configuration for your backend once more.

**Comment by halgari** [comment]
We are experiencing this same issue with dotnet 7 on ubuntu-latest. Locally and in local VMs running Ubuntu desktop LTS we do not see this issue, however in github actions, connections to or http test server (running in-process on 127.0.0.1) we see this issue on almost every run.

**Comment by vpolikarpov-akvelon** [comment]
Hey @halgari. Could you open a separate issue for that and provide more information and repro steps?

**Comment by cprice404** [comment]
@vpolikarpov-akvelon have you tried running this python reproducer on an Azure node with public IP address to see if it works in that configuration, as you said that it did with the node.js reproducer?

**Comment by vpolikarpov-akvelon** [comment]
Hey @cprice404. Did you have a chance to analyze backend logs and network security configuration? I have a strong impression that it's something out of our area of responsibility and we can't control it as there are a lot of indirect evidences that indicate that problems lays outside of runner itself.

If you have any other ideas about what we can test _within_ runners, then let me know and I'll be glad to help you with investigation.

**Comment by cprice404** [comment]
@vpolikarpov-akvelon thank you for your response, we deeply appreciate your willingness to continue checking in on this issue.

Our network security settings allow unrestricted access on 0.0.0.0 and we have never experienced this issue with any of our SDKs from outside of github, and we have never had a customer report running into anything like this. I'd be happy to share screenshots of the security group rules if that would be useful.

My teammate commented out those signing key tests in the python reproducer PR because they don't really match the normal pattern that we see on github across all of our programming languages. The tests still failed and now the error looks more like what we typically see:

https://github.com/momentohq/client-sdk-python/actions/runs/5815221224/job/15766468549?pr=357

When we get these Timeout / DEADLINE_EXCEEDED exceptions it basically always means that the request never made it to the server. I combed through the server logs from the time range of this run and I see no evidence that the requests ever got there.

I still don't understand how this could be a security configuration issue because if it was, I would expect *all* requests to fail from the problematic ubuntu 22.04 runners, as opposed to sporadic/intermittent requests. And I also would be surprised if Azure had configured different IP ranges for the 22.04 runners than the 20.04; have you found any documentation or seen any patterns that would lead you to believe that?

If you ran those python tests on an Azure runner with a public IP address and took a pcap, and we compared that to one from a failed run on a runner without a public IP address, do you think that might offer any clues?

**Comment by cprice404** [comment]
Just to clarify re: security group rules - our service is public so we do not have any CIDR range restrictions.

**Comment by cprice404** [comment]
all of the adjacent requests in the logs are successful / behaving as expected.

Are there steps documented somewhere about how to reproduce a gh actions run directly? I feel like the best next steps I can think of here are to do some runs on Azure instances covering the matrix of these dimensions:
* with and without public IPs
* with and without the gh actions environment wrapping them

If we do runs on Azure with and without public IPs, but just do them directly on the azure instance without the gh actions environment, and they are always successful, and then we do runs with and without the gh actions environment and it passes with public IPs and fails with private IPs, then that would suggest that it is something that has to do with the combo of gh actions + private IPs, would it not?

**Comment by vpolikarpov-akvelon** [comment]
Hey. Bad news. I've got a firm evidence that there is something wrong with network but not runners itself. I've tried to tunnel requests over some 3rd server and everything started working as expected.

Here is a link to the run: https://github.com/vpolikarpov-akvelon/client-sdk-python/actions/runs/5854627487. My setup is quite absurd (socks5 tunnel over ssh + local privoxy that forwards requests to tunnel) as I was building it in a rough-and-ready manner.

I understand that it doesn't help you resolving your issue. Also I understand that it's more important for you then for anyone else as this issue may not only affect your test pipelines but also pipelines of your customers. But since there is nothing wrong with runner images I don't know how could we help you resolving this.

As a workaround I may only suggest you to configure proxy as I did or to run tests on self-hosted runners.

**Comment by cprice404** [comment]
thanks for looking into it. So just to make sure I understand, you're saying that you think there's some networking issue somewhere between Azure's private-IP instances and AWS's load balancers that is causing it? And if you proxy through another host so that the networking is going from that host to AWS instead of from Azure to AWS, that avoids the networking configuration issue and resolves the problem?

**Comment by vpolikarpov-akvelon** [comment]
Yes, this is exactly what I observe. Unfortunately, we can't narrow a search and identify what node causes this issue. As tunneling requests out of GH infra solves this problem it looks logical that it's something on GH side, but I wouldn't bet on it as it may only appear when GH and AWS meet each other.

If you want to investigate it and if you have a time for it, you may try investigating how proxy server affects tests more thoroughly. In my test I used ssh tunneling from runner to 3rd server, it makes traffic not only comes to AWS from different public IP address, but also make it completely different on GH's side (ssh instead of tls), so it only proves that there is nothing wrong with runner software itself, but says nothing about GH and AWS infra. So you may try to rule them out separately in order to identify who is to blame.

Anyway, as problem lays outside runners themselves, I'm closing this thread. Feel free to open a new one or start a discussion in corresponding section if you have any other questions. Sorry for not being able to resolve your issue.

**Comment by cprice404** [comment]
@vpolikarpov-akvelon can I ask you one more question about this?

Is there something different about how gh actions provision runner instances for ubuntu 20.04 vs. 22.04? If your theory about the networking is correct then it seems like there must be a difference there, right?

**Comment by vpolikarpov-akvelon** [comment]
There is no difference in how GH provision Ubuntu runners of different versions, but of course they are grouped into some kind of pools that may 'physically' be located differently. Also please take into account that 22.04 runners match 'latest' alias currently so they are used more than 20.04, and it also may affect something.

**Comment by cprice404** [comment]
@vpolikarpov-akvelon do your Azure runners all get created in a specific region?

**Comment by cprice404** [comment]
yes, I meant GH runners. Just wondering how the region gets chosen.

Also, I set up some Azure instances today, Ubuntu 20.04 and 22.04, with and without public IP. I got the same results we've been seeing on github runners. (22.04 with private IP is the only one that demonstrated the transient failures.)

Since github has a relationship with Azure and you are effectively a customer of their VM service in this case, are you able to open a support ticket with Azure directly to make sure they are aware of this issue?

**Comment by chkimes** [comment]
I received a link to this issue and, admittedly, I have done very little investigation. However after a brief skim of the behaviors described here, it feels like this is perhaps related to SNAT exhaustion? https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections

This would explain both why private vs. public IPs make a difference, why "resting" the tests helps them succeed, and why a SOCKS5 proxy makes the problem go away (connections are made to the local proxy instead of leaving the machine).

Do the tests make a large number of outbound connections in a short period of time?

**Comment by cprice404** [comment]
Thanks a lot for that link.

The tests do make a fair number of connections, but it would be on the order of 10s to maybe a hundred. Is there something I should intuit from this table:

https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections#preallocatedports

that would give me an idea what the maximum number of ports available would be?

Also, I'm still a little lost as to why these SNAT limits would impact an Ubuntu 22.04 VM but not an Ubuntu 20.04 VM?

**Comment by chkimes** [comment]
We expect that failures shouldn't happen until the thousands of connections range. I don't have an immediate theory for why it works in one and not the other, but I would not be surprised to learn about connection management behaviors changing with different versions of underlying software (e.g. connection pooling, HTTP library connection handling).

Do we have packet captures of both succeeding and failing runs? That would help immensely in understanding any behavioral differences.

**Comment by cprice404** [comment]
Yes, we should have them from the previous incarnation of this ticket, #6981. I will ask my colleague @malandis to link them again.

**Comment by malandis** [comment]
@chkimes there are some captures on the ticket that @cprice404 mentioned.

- The failed run packet capture is [here](https://drive.google.com/file/d/1ic81EGUrhrMPLFXeofzIhZt3YKq7HJCJ/view?usp=sharing) (and for context was originally mentioned [here](https://github.com/actions/runner-images/issues/6981#issuecomment-1403893385))
- The successful run packet capture is [here](https://drive.google.com/file/d/18zgvbIEgg1K0Km5ouZp4a-8MiVGZw1yo/view?usp=sharing) (and for context was originally mentioned [here](https://github.com/actions/runner-images/issues/6981#issuecomment-1429043258))

**Comment by chkimes** [comment]
So I don't think this is SNAT exhaustion. The failed capture only shows 37 SYN packets to non-Azure-internal endpoints and the successful capture is still only 196. These are well below where I'd expect to see issues from running out of ports.

I'm having a bit of trouble mapping the packet capture to a particular failure mode. Typically I would approach this by mapping the timestamps of application behaviors in the logs to network behaviors in the packet capture, however the Actions logs attached to these packet captures have long since expired so instead I'm searching for needles in the haystack. Are there any more recent captures that I can correlate with logs that haven't expired yet?

The only thing I'm able to immediately notice here (aside from the obviously shorter pcap length on the failed job) is that the DNS result from resolved on Ubuntu 22.04 has an IP packet TTL of 1, while on 20.04 the IP packet TTL is 64. While I don't fully understand this, I also don't think it's having an impact since applications are clearly using the DNS result in later requests.

I'll dig through this capture a little more, but root causing may be difficult without the logs.

**Comment by chkimes** [comment]
Well I found a needle. However, I don't understand it.

I found some [recently failing logs](https://github.com/momentohq/client-sdk-python/actions/runs/5815221224/job/15766468549?pr=357#step:13:97) and checked the failing test, which appears to be using a timeout of 60 seconds: https://github.com/momentohq/client-sdk-python/blob/3ca90e624bb031c616ac36dbf6f85e6cd955d1c5/src/momento/internal/aio/_scs_control_client.py#L31

At 16:23:59, the test run is clearly cleaning up a bunch of connections that it is likely aborting because something else failed and the process is cleaning up:

![image](https://github.com/actions/runner-images/assets/1936066/17159471-24f2-4f5c-802e-f47c0246c721)

So I decided to check what's happening 60 seconds earlier in hopes of catching something that is starting the behavior that eventually fails.

![image](https://github.com/actions/runner-images/assets/1936066/69b6ea7e-955e-4694-bf81-e983401a6b8c)

So there are a few things that are very odd here. One is that there is a DNS query to your infrastructure that is not immediately followed by a new connection. Every other DNS query in the capture has an immediate SYN packet (because of course why make the DNS query if you're not trying to initiate a connection?). Examples:

![image](https://github.com/actions/runner-images/assets/1936066/116b0a7b-7fac-4d07-bc4e-799c9cac07a7)

![image](https://github.com/actions/runner-images/assets/1936066/d3ad4f4f-aecb-4bab-975e-3d3386268442)

![image](https://github.com/actions/runner-images/assets/1936066/7614c50c-b1d5-49c4-ab1d-03e86d03caeb)

The second oddity is that 11 seconds later exactly at 16:22:59, we get a seemingly random errant SYN/ACK packet being sent to this VM. The VM rightly sends a RST packet because it has no idea what connection the packet is in reference to. The IP is otherwise entirely unmentioned in the entire packet capture:

![image](https://github.com/actions/runner-images/assets/1936066/e9a78136-2aa9-403e-869b-ac50832984fb)

This is very unexpected. Our VMs do not allow inbound traffic, so receiving a SYN/ACK from an IP/port that we have not previously initiated a connection to is not a normal condition. The timing is also incredibly suspect as well, since it's exactly 60 seconds before the test process begins cleaning up its other connections. There are a handful of mysteries on my mind at the moment.

1. Where did this packet come from?
2. If the application thinks that it sent a request around 16:22:59, where are the packets that map to that request?
3. What is different between 20.04 and 22.04 that would cause this?

I think this behavior is a smoking gun for your issue. The trick now is identifying the trigger. Correlating logs to a packet capture would help cement the above data (since currently I'm intuiting it all from pcaps with all encrypted application data). I'll look for any notable infrastructural differences between these VM types, however my expectation is that outside the VM they should largely be configured the same.

**Comment by pratik151192** [comment]
@chkimes 

## Summary of investigations

We were able to come up with a simple reproducer that exhibited deadline exceeded (and sometimes connection reset) issues that were witnessed on the GitHub runners. 
The reproducer was implemented in the Python, JavaScript, and Java SDKs and essentially does the following:

- Creates two clients within a single process and issues Momento control plane operations (create cache, list caches, and delete cache) in a round-robin fashion over 1000 iterations.
- Another experiment was done where the reproducer was changed to have only a single client but multiple processes or instances of the program was run.

### Python SDK

#### Azure Private VM Ubuntu 22.04.3 LTS

##### 2 Clients, 1 Process

[Reproducer code](https://github.com/momentohq/client-sdk-python/blob/simple-repro/examples/repro5xx/repro.py)

- With a single client doing those operations in a loop, no issues were observed over 1000 iterations.
- With two clients, the first iteration succeeded for both client (client1 and client2) each. The second iteration for client1 just hanged for 60 seconds and started throwing `deadline exceeded`. Note that this doesn't happen 100% of the time, but more like 50% of the times.

<img width="1728" alt="Screenshot 2023-08-31 at 10 36 31 AM" src="https://github.com/momentohq/dev-eco-issue-tracker/assets/5648895/da108974-e849-4879-a404-714b27113bcc">

We also took TCP dumps during the program run, and it seems like the port which established the connection just did nothing for 60 seconds and then sent the application data. One interesting thing here is in between those 60 seconds, TCP keep alive packets were seen being sent to the destination address with successful ACK. Somewhere in the 48th second:

<img width="1685" alt="Screenshot 2023-08-31 at 10 39 20 AM" src="https://github.com/momentohq/dev-eco-issue-tracker/assets/5648895/1d698473-6397-45ae-9a57-cf0ea3ad93e1">

Similar to what you witnessed, there was a random packet sent from one of Azure IP addresses and the private VM expectedly sent back a `RST`. There was no particular correlation 60 second before and after this particular timestamp though:

<img width="1682" alt="Screenshot 2023-08-31 at 10 40 01 AM" src="https://github.com/momentohq/dev-eco-issue-tracker/assets/5648895/bcb81337-c6cd-4f26-9049-898f081a76f0">

Here's the entire TCP dump for this Python program run:

[file_result_2_clients_python.pcap.zip](https://github.com/momentohq/dev-eco-issue-tracker/files/12486702/file_result_2_clients_python.pcap.zip)

##### 1 Client, 3 processes

In this experiment, the reproducer code was tweaked to just create a single instance of the client, but 3 different processes of the program was run. After 30 seconds or so, one of the processes abruptly hung and started showing similar `Deadline exceeded` errors.

#### Azure Private VM Ubuntu 20.04.6 LTS

- We weren't able to reproduce the same on this VM. However, our GitHub runners did show timeouts for one or two out of 1000+ tests on every run which forced us to run some tests on macOS platform. 

#### Azure Public VM Ubuntu 22.04.3

- We were able to reproduce the issue once, but since then we haven't been able to reproduce again. It happened only once out of 20 or so runs. So we can probably discard this one as a false alarm or actually transient.

#### AWS EC2 Private VM Ubuntu 22.04

- No issues reproduced with 2 clients

### JavaScript SDK

[Reproducer Code](https://github.com/momentohq/client-sdk-javascript/blob/repro-js/examples/nodejs/cache/repro.ts)

#### Azure Private VM Ubuntu 22.04.3 LTS

- In JavaScript, even with one client we see the program pause abruptly for 20+ seconds or so but the reproducibility is more prominent with two clients. In one instance, even with a single client the program hung for 54 seconds before making any progress. Notice the time difference between `Creating cache: 15` and `Deleting Cache: 15`

<img width="587" alt="Screenshot 2023-08-31 at 11 14 57 AM" src="https://github.com/momentohq/dev-eco-issue-tracker/assets/5648895/510f976b-0729-456d-a61a-66b2b7368010">

- With two clients, the first iteration succeeded for both client (client1 and client2) each. The second iteration for client1 just hanged for 120 seconds, didn't make any progress, and suddenly resumed execution. Notice the timestamp on the logs in the below screenshot and the time difference between `Creating cache: 2` and `Creating cache: 3`. Interestingly, `Deleting cache: 2` was also never ever printed. Then, the program made progress for 3 more operations and again got stuck for 2 mins, and so on. 

<img width="698" alt="Screenshot 2023-08-31 at 10 56 48 AM" src="https://github.com/momentohq/dev-eco-issue-tracker/assets/5648895/0fc3c00f-5407-454b-a81f-caa23e830232">

The duration of the `pause` isn't always 120 seconds, but in every run it does pause for 30 seconds or more. The mysterious Azure VM packet appeared in this case as well:

<img width="1683" alt="Screenshot 2023-08-31 at 11 06 25 AM" src="https://github.com/momentohq/dev-eco-issue-tracker/assets/5648895/d6fcb210-9a27-4c47-8c16-18612d6ef16b">

TCP dump of a run where the program hung for 30 seconds at the same execution flow mentioned above.

[file_result_2_clients_js.pcap.zip](https://github.com/momentohq/dev-eco-issue-tracker/files/12486942/file_result_2_clients_js.pcap.zip)

#### Azure Private VM Ubuntu 20.04.6 LTS

- The behavior here is less pronounced than Ubuntu 22 but still does appear every other run. In some runs, the program progresses smoothly without any hiccups, while in other runs we do see it pause for 30 seconds or so between every 3-4 operations like in the Ubuntu 22 LTS case.

#### Azure Public VM Ubuntu 22.04.3

- No issues reproduced with 2 clients

#### AWS EC2 Private VM Ubuntu 22.04

- No issues reproduced with 2 clients

**Comment by chkimes** [comment]
> One question I have - do you have any idea what that IP address for the inbound SYN/ACK packet could be? Does it appear elsewhere in the TCP dump?

Nope, I have no clue. My immediate guess here is that this is a leaky abstraction from a bug in the networking stack somewhere, and we might be receiving a packet from the load balancer that failed NAT or IP decapsulation in some way.

Since this is appearing to me like an infrastructural bug, I'm going to get Azure involved and have the networking team look over this. 

One question I have: is this reproducible with any other endpoint? If not, do you have control over the target endpoint and are you able to take a simultaneous packet capture (from VM and from the endpoint)? It would be incredibly useful to know what is happening on the other end so we can match packets in the captures. I expect Azure will ask for the same thing so it will be good to get ahead of that.

**Comment by chkimes** [comment]
Azure incident ref: 421195254

**Comment by cprice404** [comment]
@chkimes we definitely can experiment with other endpoints and set up an environment where we can get pcaps from them. We'll have someone dig into that as soon as we can spare a bit of time for it. Thanks again for your help.

**Comment by chkimes** [comment]
Thanks! Since you've got a testing setup also figured out, could you email me at my username at github dot com with information about your Azure Subscription ID and the Resource Group + VM Names that you are using to test? This will help the networking team isolate the infrastructure involved. Ideally I could provide them some examples from GitHub Actions but we destroy those VMs immediately after every use.

Alternately I can try to set up a repro on my own long-lived VM to use for investigation with them.

**Comment by cprice404** [comment]
yep, will do

**Comment by cprice404** [comment]
also if you do want to set up a repro on your own VM, @pratik151192 has it down to a very simple program to repro it now, and we would be happy to share it.

**Comment by chkimes** [comment]
Sure! It will likely be easier to coordinate with Azure if I can locally reproduce.

**Comment by pratik151192** [comment]
@chkimes to reproduce, you'd need an auth token from our console https://console.gomomento.com/ (no credit cards needed as we have free tiers)

Once you obtain an auth token, you can pull repo/file 

https://github.com/momentohq/client-sdk-python/blob/simple-repro/examples/repro5xx/repro.py

Note that the branch where that file is located is `simple-repro` and not `main` or `master`. Once you `cd` into the directory containing `repro.py`, the following commands should work:

```
export MOMENTO_AUTH_TOKEN="token you got from console"
poetry install
poetry run python -m repro --log-level=DEBUG
```

You'll need to install `poetry` https://python-poetry.org/docs/

**Comment by pratik151192** [comment]
@chkimes Following up on this.

We were able to take simultaneous pcaps from both the client and server. We are investigating the pcaps and meanwhile sharing with you all so that we can look at it in parallel.

I have attached two zip directories, and each of them contain 2 files for client and server:

- `tcpdumps_hangs_ubuntu22_private.zip` - This has two files for client and server, and this is the dumps where we can see the hanging/timeout behavior.

- `tcpdumps_no_hangs_ubuntu22_private.zip` - This has two files for client and server, and this is the dumps where we do not see the hanging/timeout behavior. Note that on the same VM, we do observe not seeing those hangs a few times. This almost always happens during the initial run of the program, and subsequent runs start hanging.

Note that both the pcaps seem noisy with respect to the amount of TLS handshakes from different IPs. We observe the same behavior form a MacOs platform where we don't see any hangs. We are narrowing down the IPs as well to focus on Momento specific handshakes; but as I said, sharing these upfront anyway :)

[tcpdumps_hangs_ubuntu22_private.zip](https://github.com/actions/runner-images/files/12665233/tcpdumps_hangs_ubuntu22_private.zip)
[tcpdumps_no_hangs_ubuntu22_private.zip](https://github.com/actions/runner-images/files/12665234/tcpdumps_no_hangs_ubuntu22_private.zip)

**Comment by chkimes** [comment]
Thanks for those simultaneous captures! Interestingly, I'm unable to match up _any_ packets between the two captures. I see some packets with similar timing but the SEQ/ACK numbers don't line up. Do you happen to have a L7 load balancer in in front of your server?

<img width="906" alt="image" src="https://github.com/actions/runner-images/assets/1936066/bce914d7-12e9-4a6e-9c45-049fa1723228">

**Comment by cprice404** [comment]
We have AWS NLBs in front of the service hosts, so those should be layer 4. They have Client IP preservation enabled so I would expect to see the azure IPs.

Pratik is out today but we'll double check a few things on our end and confirm.

**Comment by mikhailkoliada** [comment]
Hey! I am closing the issue as stale now, feel free to reach us again if new information is available we will gladly help you.

**Comment by cprice404** [comment]
@mikhailkoliada thanks. We are definitely still impacted by this but haven't had the resources to set up additional tests and capture more tcp dumps.

**Comment by cprice404** [comment]
@mikhailkoliada @chkimes just wanted to let you all know - whatever was causing this seems to be resolved on ubuntu-24.

**Comment by jsimonek** [comment]
I think we were experiencing the same bug on Ubuntu 22 and Ubuntu 24 runners. A year ago we have upgraded to Ubuntu 22 and our end-to-end tests started to get network stability problems like read timeouts and connection resets, increasing the overall runtime of the tests more than two times and resulting in flakiness. We have debugged the problem and found that this is caused by the upgrade and just stayed on Ubuntu 20.

A few weeks ago we have upgraded to Ubuntu 24, because 20 is getting deprecated. And the test instability is back. I have prepared a minimal example that reproduces the problem reliably. Based on my investigation I think that the problem is likely in GitHub customizations of Ubuntu 22.04 and 24.04 or in GitHub networking (see section workaround).

# Our infrastructure

The test makes HTTPS connections to a service running behind an AWS ELB (Elastic Load Balancer). There are multiple ELBs, each in different availability zone. When you resolve the hostname, it resolves to multiple IPs, each one is an ELB.

# Reproduction steps

The workflow [connection-instability-minimal-test.yml](https://github.com/user-attachments/files/19469140/connection-instability-minimal-test.yml.txt) makes 150 HTTPS requests in parallel using `curl`. The workflow can be executed by anyone, the service is publicly available. These are the results I get:

| Ubuntu version | Failed requests |
|----------------|-----------------|
| 20.04          | 0               |
| 22.04          | 33              |
| 24.04          | 25              |

I have reproduced the same problem with Python, Node and `openssl s_client`, which is as low level as I can get (see [connection-instability-openssl-test.yml](https://github.com/user-attachments/files/19469139/connection-instability-openssl-test.yml.txt)). When failing there are about 20 - 40 failed requests. The failures I saw were either a timeout or connection reset:

```
OpenSSL SSL_connect: Connection reset by peer in connection to stagingapi.dnanexus.com:443 
Operation timed out after 30000 milliseconds with 0 bytes received
```

# Workaround

In Ubuntu 20, the DNS resolver `systemd resolved` picked one IP out of the list returned by DNS server and returned the same IP for some duration of time. In a newer version of systemd this [has changed](https://github.com/systemd/systemd/commit/3f0a7b3a3224681fec0133fc5ef8d30f0f9b4fa5) and in Ubuntu 22 onwards a random IP is returned from the list. This causes the parallel requests to be done against different IPs.

The workaround for Ubuntu 22 and newer is to make the IP stable, so that no parallel requests to the same hostname are made using different IPs. One way to accomplish this is to resolve the IP and put it to `/etc/hosts`. The attached workflows contain a step that does this, it also resolves the hostname every 5 minutes and updates the entry in `/etc/hosts`.

With this workaround, there are zero failed requests in all versions of Ubuntu.

# Other investigation

- The more requests are done in parallel the more failed requests there are.
- If the connection is stable or not does not depend on the endpoint of the server at all. Even with nonexistent URL, the connection instability is reproduced.
- When I tried a custom DNS server that resolved `stagingapi.dnanexus.com` to a single IP, there are zero failed requests.
- When the custom DNS server resolved `stagingapi.dnanexus.com` to one ELB IP and an IP of the server directly, there were failed requests. This means that switching between IPs is not a problem in itself)
- I reproduced the resets on many version of NodeJS randing from 8 to 21, Python versions from 3.8.18 to 3.13, OpenSSL versions from 1.1.1a to 3.2.0.
- The problem did not reproduce on Ubuntu 22.04 running as AWS EC2 instance (even though the resolver chooses a random IP for each request).
